src/main/java/com/epam/esm/dao/impl/CertificateDaoImpl.java

```diff
@@ -0,0 +1,181 @@
+import org.springframework.jdbc.support.KeyHolder;
+import org.springframework.stereotype.Repository;
+import org.springframework.transaction.annotation.Transactional;
+import org.springframework.util.CollectionUtils;
+import org.springframework.util.MultiValueMap;
+
+import java.sql.*;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static com.epam.esm.dao.RowMapProvider.*;
+import static com.epam.esm.dao.SqlProvider.*;
+
+@Repository
+@Transactional(readOnly = true)
+public class CertificateDaoImpl implements CertificateDao, RowMapper<Certificate> {
+
+    @Autowired
+    private JdbcTemplate jdbcTemplate;
+
+    @Autowired
+    private NamedParameterJdbcTemplate namedParameterJdbcTemplate;
+
+    private List<Long> getCertificateIdsByTagId(Long tagId){
+        return jdbcTemplate.query(GET_CERTIFICATES_BY_TAG_ID, new Object[] {tagId},
+                (rs, rowNum) -> rs.getLong(1));
+    }
+
+
+    @Override
+    public List<Certificate> getAll() {
+        return jdbcTemplate.query(GET_ALL_CERTIFICATES, new CertificateDaoImpl());
+
+    }
+
+    @Override
+    public List<Certificate> getByParameters(MultiValueMap<String, String> params) throws DaoException {
+        if (CollectionUtils.isEmpty(params))
+            throw new DaoException();
+        final StringBuilder stringBuilder = new StringBuilder("SELECT * FROM gift_certificate WHERE ");
+        if (Optional.ofNullable(params.get("name")).isPresent() && Optional.ofNullable(params.get("description")).isPresent()) {
+            stringBuilder
+                    .append(" name LIKE '%")
+                    .append(params.getFirst("name"))
+                    .append("%'")
+                    .append(" and ")
+                    .append("description LIKE '%")
+                    .append(params.getFirst("description"))
+                    .append("%'");
+        }
+
+        if (Optional.ofNullable(params.get("name")).isPresent()) {
+            stringBuilder
+                    .append(" name LIKE '%")
+                    .append(params.getFirst("name"))
+                    .append("%'");
+        }
+
+        if (Optional.ofNullable(params.get("description")).isPresent()) {
+            stringBuilder
+                    .append(" description LIKE '%")
+                    .append(params.getFirst("description"))
+                    .append("%'");
+        }
+        return jdbcTemplate.query(stringBuilder.toString(), new CertificateDaoImpl());
+    }
+
+    @Transactional
+    @Override
+    public Certificate add(Certificate certificate) throws DaoException {
+        KeyHolder keyHolder = new GeneratedKeyHolder();
+        try {
+            PreparedStatementCreator preparedStatementCreator = con -> {
+                PreparedStatement preparedStatement = con.prepareStatement(ADD_NEW_CERTIFICATE, PreparedStatement.RETURN_GENERATED_KEYS);
```

Comment:
It prepare to use NamedParameterStatement.
https://www.infoworld.com/article/2077706/named-parameters-for-preparedstatement.html

Reply:
Changed to NamedParameterStatement